clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -784,4 +784,39 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
         return crc.getValue();
     }
 
+    /**
+     * Read data from the channel to the given byte buffer
+     * @param channel File channel containing the data to read from
+     * @param buffer The buffer into which bytes are to be transferred
+     * @param startPosition The file position at which the transfer is to begin; must be non-negative
+     *
+     * @throws IllegalArgumentException If invalid arguments were passed
+     * @throws IOException If some other I/O error occurs
+     */
+    public static void readFullyFrom(FileChannel channel, ByteBuffer buffer, long startPosition) throws IOException {
+        if (channel == null) {
+            throw new IllegalArgumentException("Cannot read an empty file channel.");
+        }
+        if (buffer == null) {
+            throw new IllegalArgumentException("Must specify the byte buffer for the channel.");
+        }
+        if (startPosition < 0) {
+            throw new IllegalArgumentException("The file position cannot be negative.");
+        }
+        final int totalBytesToRead = buffer.capacity();
```

Comment:
This is not correct, it should be `buffer.remaining`. It seems simpler to just do `while (buffer.remaining())` with an early exit in the case read returns `-1`.